benchmarks-mobile/protonano_app/app/src/main/java/io/grpc/benchmarks/ProtobufBenchmarker.java

```diff
@@ -0,0 +1,165 @@
+            throws Exception {
+        final int serializedSize = message.getSerializedSize();
+        return benchmark("Serialize to CodedOutputByteBufferNano", message.getSerializedSize(),
+                new Action() {
+            @Override
+            public void execute() throws IOException {
+                message.writeTo(CodedOutputByteBufferNano.newInstance(new byte[serializedSize]));
+            }
+        });
+    }
+
+    public static BenchmarkResult deserializeProtobufFromByteArray(final MessageNano message)
+            throws Exception {
+        final int serializedSize = message.getSerializedSize();
+        final byte inputData[] = MessageNano.toByteArray(message);
+        return benchmark("Deserialize from byte array", serializedSize, new Action() {
+            @Override
+            public void execute() throws Exception {
+                MessageNano.mergeFrom(message.getClass().newInstance(), inputData);
+            }
+        });
+    }
+
+    public static BenchmarkResult serializeJsonToByteArray(final String jsonString, boolean gzip)
+            throws Exception {
+        final int serializedSize = jsonString.getBytes().length;
+        final JSONObject jsonObject = new JSONObject(jsonString);
+
+        if (gzip) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+            GZIPOutputStream gos = new GZIPOutputStream(bos);
+            gos.write(jsonString.getBytes());
+            gos.close();
+            bos.close();
+
+            BenchmarkResult res = benchmark("JSON serialize to byte array (gzip)", serializedSize,
+                    new Action() {
+                @Override
+                public void execute() throws IOException {
+                    ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+                    GZIPOutputStream gos = new GZIPOutputStream(bos);
+                    gos.write(jsonObject.toString().getBytes());
+                    gos.close();
+                    bos.close();
+                    bos.toByteArray();
+                }
+            });
+            res.compressedSize = bos.toByteArray().length;
+            return res;
+        } else {
+            return benchmark("JSON serialize to byte array", serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException {
+                    jsonObject.toString().getBytes();
+                }
+            });
+        }
+    }
+
+    public static BenchmarkResult deserializeJsonfromByteArray(final String jsonString,
+                                                               boolean gzip) throws Exception {
+        final int serializedSize = jsonString.getBytes().length;
+
+        if (gzip) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+            GZIPOutputStream gos = new GZIPOutputStream(bos);
```

Comment:
This is going to be tricky.  Won't real programs use the highest compression version of Gzip?  Also, real data is usually structured which makes gzip do a better job.  (and infact some people even transpose arrays-of-objects into objects-of-arrays to get better performance)

Is any of your test data representative of real data?